plenum/server/replica.py

```diff
@@ -819,27 +826,68 @@ def trackBatches(self, pp: PrePrepare, prevStateRootHash):
         self.batches[(pp.viewNo, pp.ppSeqNo)] = [pp.ledgerId, pp.discarded,
                                                  pp.ppTime, prevStateRootHash, len(pp.reqIdr)]
 
-    def send3PCBatch(self):
-        r = 0
-        for lid, q in self.requestQueues.items():
-            # TODO: make the condition more apparent
-            if len(q) >= self.config.Max3PCBatchSize or \
-                    (self.lastBatchCreated + self.config.Max3PCBatchWait < time.perf_counter() and
-                     len(q) > 0):
-                oldStateRootHash = self.stateRootHash(lid, to_str=False)
-                ppReq = self.create3PCBatch(lid)
-                if ppReq is None:
-                    continue
-                self.sendPrePrepare(ppReq)
-                if not self.isMaster:
-                    self.node.last_sent_pp_store_helper.store_last_sent_pp_seq_no(
-                        self.instId, ppReq.ppSeqNo)
-                self.trackBatches(ppReq, oldStateRootHash)
-                r += 1
-
-        if r > 0:
-            self.lastBatchCreated = time.perf_counter()
-        return r
+    def send_3pc_batch(self):
+        sent_batches = set()
+
+        # 1. send 3PC batches with requests for every ledger
+        self._send_3pc_batches_for_ledgers(sent_batches)
+
+        # 2. for every ledger we havne't just sent a 3PC batch check if it's not fresh enough,
+        # and send an empty 3PC batch to update the state if needed
+        self._send_3pc_freshness_batch(sent_batches)
+
+        # 3. update ts of last sent 3PC batch
+        sent_batches.discard(None)
+        if len(sent_batches) > 0:
+            self.lastBatchCreated = self.get_current_time()
+
+        return len(sent_batches)
+
+    def _send_3pc_batches_for_ledgers(self, sent_batches):
+        for ledger_id, q in self.requestQueues.items():
+            if len(q) == 0:
+                continue
+
+            queue_full = len(q) >= self.config.Max3PCBatchSize
+            timeout = self.lastBatchCreated + self.config.Max3PCBatchWait < self.get_current_time()
+            if not queue_full and not timeout:
+                continue
+
+            sent_batches.add(
```

Comment:
Yes. The current logic (that I haven't changed yet) is that we may send a PrePrepare (3PC batch) for every ledger at the same time. The same for freshness.
In future it may be better to make sure that every 3PC batch is sent not earlier than X ms, but it requires a better tracing of what ledgers we sent a 3PC batches for, and some load balancing to make sure that all ledgers are being ordered.
There is a TODO comment for this.